[USDU-349] URP import support

vickycl opened the pull request:
## Purpose of this PR

**Ticket/Jira #:** [USDU-349](https://jira.unity3d.com/browse/USDU-349) / [USDU-324](https://jira.unity3d.com/browse/USDU-324) / [USDU-348](https://jira.unity3d.com/browse/USDU-348)

<!-- Description of feature/change. Links to screenshots, design docs, user docs, etc. Remember reviewers may be outside your team, and not know your feature/area that should be explained more. -->

Adds support for importing materials into a URP-based project, heavily based on the HDRP implementation. Supports basic materials and maps their inputs to URP's Lit shader. There is no dedicated implementation for export, but there is a fix for occlusion maps and specular workflows that should fix most cases of URP export through the standard shader export flow.

## Testing

**Functional Testing status:**

<!-- Explanation of what's tested, how tested and existing or new automation tests. Can include manual testing by self and/or QA. Specify test plans. Rarely acceptable to have no testing.  -->

@lee-aandrew testing results: https://docs.google.com/document/d/1wO_Hu9cORmNgmfp-dExHZ--c_qXVyONC/edit

**Performance Testing status:**

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->

Combining the smoothness into the alpha channel of the metallic map is slow, but as import tin URP wasn't available before there's nothing to baseline against.

## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:** Medium
<!-- (Minimal / Low / Medium / High) -->

**Halo Effect:** Medium (some changes to standard shader exporter)
<!-- (Minimal / Low / Medium / High) -->

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [x] Add entry in CHANGELOG.md _(if applicable)_

vickycl committed:
Initial experiment of URP support

vickycl committed:
Added todo to remove duplicated method

vickycl committed:
Merge branch 'dev' into exp/urp-import-support

vickycl committed:
Cleaned up unused code

SimonBoorer commented:
The USD Importer package supports importing materials for URP projects, no? Do they have support for mapping smoothness/roughness we can leverage?

vickycl committed:
Temporary URP export fix.

I've patched up the standard shader exporter to fix a bug where URP exported materials didn't look correct. This isn't a fully featured URP export but should plug a hole for now.

vickycl committed:
Fix for smoothness mapping.

vickycl committed:
Add note about smoothness constant value.

vickycl committed:
Set the correct shader keyword and remove smoothness from the docs limitations.

vickycl committed:
Fix occlusion mapping on export.

vickycl committed:
Small refactor to remove redundant code.

SimonBoorer commented on package/com.unity.formats.usd/CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## [3.0.0-exp.5] - Unreleased
 ### Features
+- Added basic URP Import support.
```
do I take your changes to mean that there is also basic URP export support now too?

lee-aandrew replied:
@SimonBoorer  i need to test this!

vickycl replied:
@SimonBoorer well, yes, but it's not dedicated URP export (like HDRP has), I just fixed the default shader export path to correctly export URP shaders too. So I wasn't sure if it counted as a feature 😅 I'll add it as a bug fix instead

michaeljblain commented on README.md:
```diff
@@ -63,7 +63,7 @@ We do not currently support the following:
     * Materials:
         * Transparency Settings (these can be set manually)
-        * Some values are unassigned when importing URP
+        * URP export not fully featured
```
I think we should add a bit more detail so that users can make a more informed decision about using the package with URP. Specifically I'd say we include what we know is working to the best of our knowledge.

vickycl replied:
I'm actually not sure what isn't working outside of transparency. This is more of a catch all for anything users might find to be missing due to URP just reusing the standard shader export :P

michaeljblain replied:
Ok and I see transparency is already covered. It feels like there's more work involved to narrow this down. If more specific feature gaps are reported going forward then we can populate them here.

vickycl committed:
Merge branch 'dev' into exp/urp-import-support

vickycl committed:
Added URP export to changelog.

vickycl committed:
Merge branch 'exp/urp-import-support' of https://github.com/Unity-Technologies/usd-unity-sdk into exp/urp-import-support

michaeljblain commented on package/com.unity.formats.usd/CHANGELOG.md:
```diff
@@ -2,9 +2,11 @@
 
 ### Bug Fixes
 - Fixed export for Prefabs from the Project Window.
+- Fixed export of URP base and occlusion maps.
```
I think we should include the URP shader type here e.g. Lit

vickycl replied:
👍

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Materials/StandardShaderExporter.cs:
```diff
@@ -419,7 +415,7 @@ public class StandardShaderExporter : ShaderExporterBase
             {
                 var newTex = SetupTexture(scene, usdShaderPath, mat, surface, Vector4.one, destTexturePath,
-                    "_OcclusionMap", "r");
+                    "_OcclusionMap", "g");
```
Can we add a comment here (and a link to docs) about why we use the "g" channel for occlusion.

vickycl replied:
done

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Materials/UrpShaderImporter.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+            else
+            {
+                // TODO: In URP we combine a smoothness map with a constant smoothness value, so we should actually
```
Do we already have a ticket for this?

vickycl replied:
added https://jira.unity3d.com/browse/USDU-474

michaeljblain reviewed (changes requested):
Some minor suggestions. You may also want to add all relevant jira tickets to PR description. 
Otherwise looks good. Let me know when you have next batch of commits ready.

vickycl committed:
Updates from review.

vickycl committed:
Update README.md

vickycl committed:
Update README.md

michaeljblain reviewed (approved):
Barring any further feedback from QA and pending. reviews this looks like it's in a good place.

vickycl commented:
@lee-aandrew are you doing any more testing of this or are you all done?
